### PR TITLE
Add `Performance/Caller` cop

### PR DESCRIPTION
**Pull request by alpaca-tc**
## Example

```
# bad
caller.first

# good
caller(1..1).first
```

## Benchmark

```
require 'benchmark'

MAX = 10_000

def record(count = 100, &block)
  count.zero? ? yield : record(count - 1, &block)
end

def report(benchmark, name, &block)
  benchmark.report(name) do
    record { MAX.times(&block) }
  end
end

Benchmark.bm do |x|
  report(x, 'caller.first')       { caller.first }
  report(x, 'caller(1..1).first') { caller(1, 1).first }
end

#=>
#        user     system      total        real
# caller.first  1.470000   0.020000   1.490000 (  1.515402)
# caller(1..1).first  0.030000   0.010000   0.040000 (  0.035801)
```

-----------------

Before submitting the PR make sure the following are checked:

* [x] Wrote [good commit messages][1].
* [ ] Commit message starts with `[Fix #issue-number]` (if the related issue exists).
* [x] Used the same coding conventions as the rest of the project.
* [x] Feature branch is up-to-date with `master` (if not - rebase it).
* [x] Squashed related commits together.
* [x] Added tests.
* [x] Added an entry to the [Changelog](../blob/master/CHANGELOG.md) if the new code introduces user-observable changes. See [changelog entry format](../blob/master/CONTRIBUTING.md#changelog-entry-format).
* [x] All tests are passing.
* [x] The new code doesn't generate RuboCop offenses.
* [x] The PR relates to *only* one subject with a clear title
  and description in grammatically correct, complete sentences.
* [x] Updated cop documentation with `rake generate_cops_documentation` (required only when you've added a new cop or changed the configuration/documentation of an existing cop).

[1]: http://chris.beams.io/posts/git-commit/

### CODE REVIEW

File: config/enabled.yml [hunk 1]
```diff
@@ -1342,6 +1342,11 @@ Lint/Void:
 
 #################### Performance ###########################
 
+Performance/CallerLocations:
+  Description: >-
+             Use `caller_locations` instance of `caller`
```
Comment (Drenmi) [inline config/enabled.yml]: Did you mean **instead** of? 🙂

File: lib/rubocop/cop/performance/caller_locations.rb [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+      #   caller.first
+      #
+      #   # good
+      #   caller_locations(1, 1).first.to_s
+      class CallerLocations < Cop
+        MSG = 'Use `caller_locations` instead of `caller`.'.freeze
+        SCOPE_METHODS = [:first, :[]].freeze
+
+        def on_send(node)
+          add_offense(node, node.loc.selector, MSG) if unneeded_caller?(node)
```
Comment (Drenmi) [inline lib/rubocop/cop/performance/caller_locations.rb]: There should be a shortcut that allows you to do this:

```ruby
add_offense(node, :selector)
```

File: spec/rubocop/cop/performance/caller_locations_spec.rb [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+  subject(:cop) { described_class.new }
+
+  it "doesn't register an offense when caller is called" do
+    inspect_source(cop, 'caller')
+    expect(cop.messages).to be_empty
+  end
+
+  it 'registers an offense when :[] is called on caller' do
+    inspect_source(cop, 'caller[0]')
+    expect(cop.messages).to eq([described_class::MSG])
```
Comment (Drenmi) [inline spec/rubocop/cop/performance/caller_locations_spec.rb]: We would normally assert:

```ruby
expect(cop.offenses.size).to eq(1)
```

Asserting messages is generally only done for cops that have messages that change depending on context. 🙂

File: config/enabled.yml (hunk 1 above)
Comment (alpaca-tc) [inline config/enabled.yml]: ooops... 😇

**Comment by alpaca-tc** [comment]
Rails4.2.8なら通りそうな気がする... 😇

File: spec/rubocop/cop/performance/caller_locations_spec.rb [hunk 4]
```diff
@@ -0,0 +1,22 @@
+# frozen_string_literal: true
+
+require 'spec_helper'
+
+fdescribe RuboCop::Cop::Performance::CallerLocations do
```
Comment (sachin21) [inline spec/rubocop/cop/performance/caller_locations_spec.rb]: This look like filter option.
If are you using filter option in development, you have to remove after development.

**Comment by bbatsov** [comment]
Some more background on this would be useful as the replacement code seems way more complex than the straightforward "slow" code.

**Comment by Dorian** [comment]
@alpaca-tc Seems like it's fixable (lazy enumerator instead of Array, lazy hash, lazy array, etc) (function is `vm_backtrace_to_ary`).

Also, `caller(1..1)` would be a good compromise IMHO (https://ruby-doc.org/core-2.4.0/Kernel.html#method-i-caller_locations)

**Comment by alpaca-tc** [comment]
@bbatsov I fixed to simply replace with `caller(1..1)` rather than complex `caller_locations`.

@Dorian Thank you for reviewing 👍 

> Seems like it's fixable (lazy enumerator instead of Array, lazy hash, lazy array, etc) (function is vm_backtrace_to_ary).

hmm, could you tell me more details to what should i fix? 🤔

**Comment by bbatsov** [comment]
Just two more things:

* in the first commit message drop the final `.`
* in the second commit message - replace `to` with `with`

**Comment by bbatsov** [comment]
Btw, shouldn't this cop be more generic? If you do `caller[n]` it should be replaced with `caller(n..n)`? Seems a bit odd to optimize just for one case.

**Comment by bbatsov** [comment]
Any progress here? If not - I'll be forced to close this due to lack of activity.

**Comment by alpaca-tc** [comment]
Sorry, I kept you waiting. I try to improve this pr :pray:

**Comment by rrosenblum** [comment]
Just adding a bit more bench-marking information because I was curious.

```ruby
[14] pry(main)> Benchmark.ips do |x|
[14] pry(main)*   x.report('caller.first') { caller.first }  
[14] pry(main)*   x.report('caller(1..1).first') { caller(1..1).first }  
[14] pry(main)*   x.compare!
[14] pry(main)* end  
Warming up --------------------------------------
        caller.first     2.314k i/100ms
  caller(1..1).first    36.661k i/100ms
Calculating -------------------------------------
        caller.first     24.241k (±13.5%) i/s -    120.328k in   5.055955s
  caller(1..1).first    490.166k (±16.5%) i/s -      2.383M in   5.028717s

Comparison:
  caller(1..1).first:   490166.1 i/s
        caller.first:    24240.7 i/s - 20.22x  slower
```

Most of the ideas for our performance cops came from [fast-ruby](https://github.com/JuanitoFatas/fast-ruby). I didn't see this one listed. It might be nice to make a PR against their project as well.

**Comment by alpaca-tc** [comment]
@bbatsov Fixed 👍

**Comment by sachin21** [comment]
@alpaca-tc You have to rebase and resolve conflicts.

**Comment by bbatsov** [comment]
The messages at least still mention `1` explicitly instead of being a generic index. Perhaps you forgot about them? I also don't see any tests for anything but `1`.

File: lib/rubocop/cop/performance/caller.rb [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+          add_offense(node, :selector)
+        end
+
+        private
+
+        def slow_caller?(node)
+          arguments = node.receiver.arguments
+
+          arguments.empty? ||
+            (arguments.length == 1 && arguments[0].type == :int)
```
Comment (rrosenblum) [inline lib/rubocop/cop/performance/caller.rb]: Should this check `arguments[0].int_type?` instead of `arguments[0].type == :int`?

Comment (bbatsov) [inline lib/rubocop/cop/performance/caller.rb]: Yes.

**Comment by bbatsov** [comment]
You should rebase this on top of the current `master`.

File: lib/rubocop/cop/performance/caller.rb [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+      # @example
+      #   # bad
+      #   caller[n]
+      #   caller.first
+      #
+      #   # good
+      #   caller(n..n).first
+      #   caller(1..1).first
+      class Caller < Cop
+        MSG = 'Use `caller(n..n)` instead of `caller`.'.freeze
```
Comment (bbatsov) [inline lib/rubocop/cop/performance/caller.rb]: I guess instead of `caller[n]` would be more clear here.

**Comment by aripollak** [comment]
@alpaca-tc how come this suggests `1..1` instead of `1, 1`? Either one has the same number of characters, but the former doesn't also require creating a Range object.

**Comment by alpaca-tc** [comment]
@aripollak yeah, it sounds nice

**Comment by rrosenblum** [comment]
I too noticed that the message could use some clarification. "Use `caller(n..n)` instead of `caller[n]`" isn't fully accurate. Wouldn't the correction need to be `caller(n..n)[n]`?

**Pull request merged**